daemons/ipa-otpd/passkey.c

```diff
@@ -0,0 +1,825 @@
+        free(data);
+        return NULL;
+    }
+
+    c = 0;
+    idx = 0;
+    do {
+        tmp = krad_packet_get_attr(req, attr, c);
+        if (tmp != NULL) {
+            memcpy(data->data + idx, tmp->data, tmp->length);
+            idx += tmp->length;
+        }
+        c++;
+    } while (tmp != NULL);
+
+    if (idx != data->length) {
+        free(data->data);
+        free(data);
+        return NULL;
+    }
+
+    return data;
+}
+
+/* Parse the passkey configuration */
+const char *otpd_parse_passkey(LDAP *ldp, LDAPMessage *entry,
+                               struct otpd_queue_item *item)
+{
+    int i;
+    char **objectclasses = NULL;
+
+    if (item->passkey == NULL) {
+        otpd_log_req(item->req,
+                     "Missing passkey struct to store passkey configuration");
+        return strerror(EINVAL);
+    }
+
+    while (entry != NULL) {
+        i = get_string_array(ldp, entry, "objectclass", &objectclasses);
+        if (i != 0) {
+            return strerror(i);
+        }
+
+        if (auth_type_is(objectclasses, "ipapasskeyconfigobject")) {
+            free(objectclasses);
+
+            i = get_string(ldp, entry, "ipaRequireUserVerification",
+                           &item->passkey->ipaRequireUserVerification);
+            if ((i != 0) && (i != ENOENT)) {
+                return strerror(i);
+            }
+        } else if (auth_type_is(objectclasses, "domainRelatedObject")) {
+            free(objectclasses);
+
+            i = get_string(ldp, entry, "associatedDomain",
+                           &item->passkey->domain);
+            if ((i != 0) && (i != ENOENT)) {
+                return strerror(i);
+            }
+        }
+
+        entry = ldap_next_entry(ldp, entry);
+    };
+
+    return NULL;
+}
+
+static int decode_json(const char *inp, size_t size, struct passkey_data *data)
+{
+    json_error_t jret;
+    int ret;
+
+    data->jroot = json_loadb(inp, size, 0, &jret);
+    if (data->jroot == NULL) {
+        return EINVAL;
+    }
+    data->jdata = NULL;
+    data->phase = -1;
+
+    ret = json_unpack(data->jroot, "{s:i, s?:s, s?:o}",
+                     "phase", &data->phase,
+                     "state", &data->state,
+                     "data", &data->jdata);
+    if (ret != 0) {
+        ret = EINVAL;
+        goto done;
+    }
+
+    switch (data->phase) {
+    case 0: /* SSS_PASSKEY_PHASE_INIT */
+        /* no data */
+        if (data->jdata != NULL) {
+            ret = EINVAL;
+        } else {
+            ret = 0;
+        }
+        break;
+    case 2: /* SSS_PASSKEY_PHASE_REPLY */
+        ret = json_unpack(data->jdata, "{s:s, s:s, s:s, s:s}",
+                "credential_id", &data->data.response.credential_id,
+                "cryptographic_challenge", &data->data.response.cryptographic_challenge,
+                "authenticator_data", &data->data.response.authenticator_data,
+                "assertion_signature", &data->data.response.assertion_signature);
```

Comment:
The reply now contains a user_id field, does it need to be added here? (I see the passkey krb5 plugin does account for this)
~~~
   "data" : {
      "assertion_signature" : "MEQCIGJnqBgzsqilI5aw/eQBVJIZ+/z7K2OiCvLEeLxD4lybAiBzPNoyswh9K3/G9vhxG+UWe9IPsrrPlPR6/HrZCcCudA==",
      "authenticator_data" : "WCWcAO58pq6rRNqWEXg/P4ZkES9OqDS8L29D9KqffS71WwUAAAFP",
      "credential_id" : "xg6rAN18gKM/iI+OFj3cdEs9Fl9kw6MbHt7L+w5ntLk96MVtLUBmDEUSq2gRfD7X4MsDnercMW93xurY0yRyZA==",
      "cryptographic_challenge" : "GLtKBNjC0mT2yTBdqtcZQ2hGMQkDlCLiuLIpxwe3jms=",
      "user_id" : ""
   },

Reply:
Yes, the `user_id` field should also be taken into account in the plugin